### DISCUSSION TITLE
eval-string in a jpm binary

### DISCUSSION THREAD

**Discussion by evanfarrar** [janet-lang/janet#1464] [Q&A]
Can anyone explain to me why `eval-string` gives me an error `error: unknown symbol hello-world` when running this jpm built executable? 

```init.janet
(def hello-world "heyheyhey")
(defn main [&] (print (eval-string "hello-world")))
```
```project.janet 
(declare-project :name "janettest")
(declare-executable :name "janettest" :entry "init.janet")
```

I can get it to work in a REPL, but not in a jpm executable. At first I thought the unused symbols were dropped out by tree-shaking, but I do see "heyheyhey" in the generated c code; I've tried a few other ways to ensure the symbol is available at runtime (such as defining it in the main function or explicitly referencing it) but no such luck. I don't know that I need it work, but I'm curious to understand it. Metaprogramming and compilation to compact native binaries are chocolate and peanut butter, I can't imagine not trying to use two great things together in the future.

**Comment by sogaiu** [comment]
Does this help?

```janet
(def hello-world "heyheyhey")

(def ce (curenv))

(defn main [&]
  (print (eval-string "hello-world" ce)))
```

I think the above was inspired by the coverage of compilation in [this chapter](https://janet.guide/compilation-and-imagination/) of @ianthehenry's book.  At least, that's what my subconscious seems to be saying (^^;

Perhaps there is a better way though.

---

Note that [`eval-string` got an extra argument recently](https://github.com/janet-lang/janet/releases/tag/v1.35.0), so the above may not work in earlier setups.
